File: native-engine/blaze-cudf-bridge/bridge-cpp/src/expr.cpp
// Copyright 2022 The Blaze Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "expr.hpp"
#include <cudf/ast/expressions.hpp>
#include <cudf/scalar/scalar.hpp>
#include <memory>
#include <stdexcept>

size_t ExprBuilder::add_column_reference(size_t column_idx) {
    auto expr_id = tree.size();
    tree.emplace<cudf::ast::column_reference>(column_idx);
    return expr_id;
}

size_t ExprBuilder::add_column_name_reference(std::string const& name) {
    auto expr_id = tree.size();
    tree.emplace<cudf::ast::column_name_reference>(name);
    return expr_id;
}

size_t ExprBuilder::add_unary_operation(size_t arg1_expr_id, int op) {
    auto const& arg1 = tree.at(arg1_expr_id);
    auto expr_id = tree.size();
    tree.emplace<cudf::ast::operation>((cudf::ast::ast_operator)op, arg1);
    return expr_id;
}

size_t ExprBuilder::add_binary_operation(size_t arg1_expr_id, size_t arg2_expr_id, int op) {
    auto const& arg1 = tree.at(arg1_expr_id);
    auto const& arg2 = tree.at(arg2_expr_id);
    auto expr_id = tree.size();
    tree.emplace<cudf::ast::operation>((cudf::ast::ast_operator)op, arg1, arg2);
    return expr_id;
}

size_t ExprBuilder::add_literal(std::shared_ptr<cudf::scalar> scalar) {
    auto& scalar_value = *scalar.get();
    auto&& type_id = typeid(scalar_value);
    auto expr_id = tree.size();

#define handle_numeric_type(c_ty)                                                                  \
    if (type_id == typeid(cudf::numeric_scalar<c_ty>)) {                                           \
        tree.emplace<cudf::ast::literal>(dynamic_cast<cudf::numeric_scalar<c_ty>&>(scalar_value)); \
        scalar_pool.push_back((scalar));                                                           \
        return expr_id;                                                                            \
    }
    handle_numeric_type(int8_t);
    handle_numeric_type(int16_t);
    handle_numeric_type(int32_t);
    handle_numeric_type(int64_t);
    handle_numeric_type(uint8_t);
    handle_numeric_type(uint16_t);
    handle_numeric_type(uint32_t);
    handle_numeric_type(uint64_t);
    handle_numeric_type(float);
    handle_numeric_type(double);

    if (type_id == typeid(cudf::string_scalar)) {
        tree.emplace<cudf::ast::literal>(dynamic_cast<cudf::string_scalar&>(scalar_value));
        scalar_pool.push_back(scalar);
        return expr_id;
    }
    throw std::runtime_error(std::string("unsupported scalar type: ") + type_id.name());
}

cudf::ast::expression const& ExprBuilder::expr() const {
    return tree.back();
}

File: native-engine/datafusion-ext-plans/src/cudf/expr.rs
// Copyright 2022 The Blaze Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use blaze_cudf_bridge::exprs::{CudfExprBuilder, CudfExprOperator, CudfScalar};
use datafusion::{
    error::Result,
    logical_expr::Operator,
    physical_expr::PhysicalExprRef,
    physical_plan::expressions::{
        BinaryExpr, Column, IsNotNullExpr, IsNullExpr, Literal, NegativeExpr,
    },
    scalar::ScalarValue,
};
use datafusion_ext_commons::{df_unimplemented_err, downcast_any};

pub fn convert_datafusion_expr_to_cudf(
    de: &PhysicalExprRef,
    use_column_name: bool,
) -> Result<CudfExprBuilder> {
    let root_expr_builder = CudfExprBuilder::new();
    let _ = convert_impl(&de, &root_expr_builder, use_column_name)?;
    Ok(root_expr_builder)
}

fn convert_impl(de: &PhysicalExprRef, b: &CudfExprBuilder, use_column_name: bool) -> Result<usize> {
    if let Ok(lit) = downcast_any!(de, Literal) {
        let cudf_scalar = match lit.value() {
            ScalarValue::Boolean(Some(v)) => CudfScalar::new_bool(*v),
            ScalarValue::Int8(Some(v)) => CudfScalar::new_prim(*v),
            ScalarValue::Int16(Some(v)) => CudfScalar::new_prim(*v),
            ScalarValue::Int32(Some(v)) => CudfScalar::new_prim(*v),
            ScalarValue::Int64(Some(v)) => CudfScalar::new_prim(*v),
            ScalarValue::UInt8(Some(v)) => CudfScalar::new_prim(*v),
            ScalarValue::UInt16(Some(v)) => CudfScalar::new_prim(*v),
            ScalarValue::UInt32(Some(v)) => CudfScalar::new_prim(*v),
            ScalarValue::UInt64(Some(v)) => CudfScalar::new_prim(*v),
            ScalarValue::Float32(Some(v)) => CudfScalar::new_prim(*v),
            ScalarValue::Float64(Some(v)) => CudfScalar::new_prim(*v),
            ScalarValue::Utf8(Some(v)) => CudfScalar::new_string(v),
            _ => return df_unimplemented_err!("cannot convert scalar to cudf: {lit:?}"),
        };
        return Ok(b.add_literal(&cudf_scalar));
    } else if let Ok(col) = downcast_any!(de, Column) {
        if use_column_name {
            return Ok(b.add_column_name_ref(col.name()));
        } else {
            return Ok(b.add_column_ref(col.index()));
        }
    } else if let Ok(neg) = downcast_any!(de, NegativeExpr) {
        let arg_expr_id = convert_impl(neg.arg(), b, use_column_name)?;
        return Ok(b.add_unary(arg_expr_id, CudfExprOperator::SUB));
    } else if let Ok(binary) = downcast_any!(de, BinaryExpr) {
        let left_expr_id = convert_impl(binary.left(), b, use_column_name)?;
        let right_expr_id = convert_impl(binary.right(), b, use_column_name)?;
        let op = convert_op(binary.op())?;
        return Ok(b.add_binary(left_expr_id, right_expr_id, op));
    } else if let Ok(is_null) = downcast_any!(de, IsNullExpr) {
        let arg_expr_id = convert_impl(is_null.arg(), b, use_column_name)?;
        return Ok(b.add_unary(arg_expr_id, CudfExprOperator::IS_NULL));
    } else if let Ok(is_not_null) = downcast_any!(de, IsNotNullExpr) {
        let arg_expr_id = convert_impl(is_not_null.arg(), b, use_column_name)?;
        let is_null_expr_id = b.add_unary(arg_expr_id, CudfExprOperator::IS_NULL);
        return Ok(b.add_unary(is_null_expr_id, CudfExprOperator::NOT));
    }
    df_unimplemented_err!("cannot convert expr to cudf: {de:?}")
}

fn convert_op(op: &Operator) -> Result<CudfExprOperator> {
    Ok(match op {
        Operator::Eq => CudfExprOperator::EQUAL,
        Operator::NotEq => CudfExprOperator::NOT_EQUAL,
        Operator::Lt => CudfExprOperator::LESS,
        Operator::LtEq => CudfExprOperator::LESS_EQUAL,
        Operator::Gt => CudfExprOperator::GREATER,
        Operator::GtEq => CudfExprOperator::GREATER_EQUAL,
        Operator::Plus => CudfExprOperator::ADD,
        Operator::Minus => CudfExprOperator::SUB,
        Operator::Multiply => CudfExprOperator::MUL,
        Operator::Divide => CudfExprOperator::DIV,
        Operator::Modulo => CudfExprOperator::MOD,
        Operator::And => CudfExprOperator::LOGICAL_AND,
        Operator::Or => CudfExprOperator::LOGICAL_OR,
        Operator::BitwiseAnd => CudfExprOperator::BITWISE_AND,
        Operator::BitwiseOr => CudfExprOperator::BITWISE_OR,
        Operator::BitwiseXor => CudfExprOperator::BITWISE_XOR,
        other => return df_unimplemented_err!("connot convert operator to cudf: {other:?}"),
    })
}
